File: level3/섬연결하기.py
import heapq

def Union(parent, node1, node2):
    p1 = Find(parent, node1)
    p2 = Find(parent, node2)
    parent[p2] = p1        
    return parent

def Find(parent, node):
    while parent[node] != node:
        node = parent[node]
    return node

def solution(n, costs):
    q = []
    for n1, n2, c in costs:
        heapq.heappush(q, (c,n1,n2))
    
    answer = 0
    parent = [i for i in range(n)]
    while q:
        c, n1, n2 = heapq.heappop(q)
        if Find(parent,n1) == Find(parent,n2):
            continue
            
        parent = Union(parent,n1,n2)
        answer += c
        
    return answer

File: level3/섬연결하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

struct Island{
    int node, cost;
};

struct cmp{
    bool operator()(Island &a, Island &b){
        if(a.cost == b.cost){
            return a.node > b.node;
        }
        return a.cost > b.cost;
    }
};

int solution(int n, vector<vector<int>> costs) {
    vector<vector<int>> cost(n, vector<int>(n,987654321));
    for(int i = 0; i < costs.size(); i++){
        int s = costs[i][0], e = costs[i][1], c = costs[i][2];
        cost[s][e] = c;
        cost[e][s] = c;
    }   
    
    priority_queue<Island,vector<Island>,cmp> pq;
    pq.push({0,0});
    for(int i = 1; i < n; i++){
        pq.push({i,987654321});
    }
    
    int answer = 0;
    while(!pq.empty()){
        int node = pq.top().node, n_cost = pq.top().cost;
        pq.pop();
        answer += n_cost;
        
        priority_queue<Island,vector<Island>,cmp> tmp;
        while(!pq.empty()){
            int tmp_node = pq.top().node, tmp_cost = pq.top().cost;
            pq.pop();
            
            if(cost[node][tmp_node] < tmp_cost){
                tmp.push({tmp_node,cost[node][tmp_node]});
            }
            else{
                tmp.push({tmp_node,tmp_cost});
            }
        }
        pq = tmp;
    }
    
    return answer;
}
